src/chainstate/burn/operations/peg_in.rs

```diff
@@ -0,0 +1,470 @@
+use stacks_common::codec::StacksMessageCodec;
+
+use crate::burnchains::BurnchainBlockHeader;
+use crate::burnchains::BurnchainTransaction;
+use crate::chainstate::burn::Opcodes;
+use crate::types::chainstate::StacksAddress;
+use crate::types::Address;
+
+use crate::chainstate::burn::operations::Error as OpError;
+use crate::chainstate::burn::operations::PegInOp;
+
+use crate::vm::errors::RuntimeErrorType as ClarityRuntimeError;
+use crate::vm::types::PrincipalData;
+use crate::vm::types::QualifiedContractIdentifier;
+use crate::vm::types::StandardPrincipalData;
+use crate::vm::ContractName;
+
+impl PegInOp {
+    pub fn from_tx(
+        block_header: &BurnchainBlockHeader,
+        tx: &BurnchainTransaction,
+    ) -> Result<Self, OpError> {
+        if tx.opcode() != Opcodes::PegIn as u8 {
+            warn!("Invalid tx: invalid opcode {}", tx.opcode());
+            return Err(OpError::InvalidInput);
+        }
+
+        let (amount, peg_wallet_address) =
+            if let Some(Some(recipient)) = tx.get_recipients().first() {
+                (recipient.amount, recipient.address.clone())
+            } else {
+                warn!("Invalid tx: Output 2 not provided");
+                return Err(OpError::InvalidInput);
+            };
+
+        let parsed_data = Self::parse_data(&tx.data())?;
+
+        let txid = tx.txid();
+        let vtxindex = tx.vtxindex();
+        let block_height = block_header.block_height;
+        let burn_header_hash = block_header.block_hash;
+
+        Ok(Self {
+            recipient: parsed_data.recipient,
+            peg_wallet_address,
+            amount,
+            memo: parsed_data.memo,
+            txid,
+            vtxindex,
+            block_height,
+            burn_header_hash,
+        })
+    }
+
+    fn parse_data(data: &[u8]) -> Result<ParsedData, ParseError> {
+        /*
+            Wire format:
+
+            0      2  3                  24                            64       80
+            |------|--|------------------|-----------------------------|--------|
+             magic  op   Stacks address      Contract name (optional)     memo
+
+             Note that `data` is missing the first 3 bytes -- the magic and op must
+             be stripped before this method is called. At the time of writing,
+             this is done in `burnchains::bitcoin::blocks::BitcoinBlockParser::parse_data`.
+        */
+
+        if data.len() < 21 {
+            warn!(
+                "PegInOp payload is malformed ({} bytes, expected at least {})",
+                data.len(),
+                21
+            );
+            return Err(ParseError::MalformedData);
+        }
+
+        let version = *data.get(0).ok_or(ParseError::MalformedData)?;
```

Comment:
Because we know that `data.len() >= 21`, you can instead use `.expect()` here instead of `.ok_or()`.

Reply:
Done.